NPE Comparing alternative priorities

**Describe the bug**
`Beans.compareAlternativeBeans` always prefers the value of the declaring bean, even if the declaring bean doesn't have an alternative priority and the bean itself does. This leads to an NPE when comparing alternate priorities.

**Expected behavior**
I expect something like this to work be resolvable (not sure this exactly reproduces.. I'm working on it)

```
    @Dependent
    static class NoParentAlternativePriority {

        @Produces
        @ApplicationScoped
        @AlternativePriority(3)
        PrioritizedProducerMethodBean bar = new PrioritizedProducerMethodBean();

        @Produces
        @ApplicationScoped
        @AlternativePriority(4)
        public PrioritizedProducerMethodBean createBar() {
            return new PrioritizedProducerMethodBean();
        }
    }
```

**Actual behavior**
As it happened in my situation, the comparison resulted in an NPE: 

```[ERROR] dev.ebullient.micrometer.deployment.export.SecondPrometheusTest  Time elapsed: 0.354 s  <<< ERROR!
java.lang.RuntimeException: 
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: java.lang.NullPointerException
        at io.quarkus.arc.processor.Beans.compareAlternativeBeans(Beans.java:551)
        at java.base/java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
        at java.base/java.util.TimSort.sort(TimSort.java:220)
        at java.base/java.util.Arrays.sort(Arrays.java:1515)
        at java.base/java.util.ArrayList.sort(ArrayList.java:1749)
        at io.quarkus.arc.processor.Beans.resolveAmbiguity(Beans.java:523)
        at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:475)
        at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:420)
        at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:224)
        at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:130)
        at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:291)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:938)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:273)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

Will create a PR to fix the problem that includes a reproducing test.

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
@@ -558,10 +558,24 @@ final class Beans {
 
     private static int compareAlternativeBeans(BeanInfo bean1, BeanInfo bean2) {
         // The highest priority wins
-        Integer priority2 = bean2.getDeclaringBean() != null ? bean2.getDeclaringBean().getAlternativePriority()
-                : bean2.getAlternativePriority();
-        Integer priority1 = bean1.getDeclaringBean() != null ? bean1.getDeclaringBean().getAlternativePriority()
-                : bean1.getAlternativePriority();
+        Integer priority1, priority2;
+
+        priority2 = bean2.getAlternativePriority();
+        if (priority2 == null) {
+            priority2 = bean2.getDeclaringBean().getAlternativePriority();
+        }
+
+        priority1 = bean1.getAlternativePriority();
+        if (priority1 == null) {
+            priority1 = bean1.getDeclaringBean().getAlternativePriority();
+        }
+
+        if (priority2 == null) {
+            return priority1 == null ? 0 : -1;
+        } else if (priority1 == null) {
+            return 1;
+        }
+
         return priority2.compareTo(priority1);
     }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
@@ -570,10 +570,11 @@ final class Beans {
             priority1 = bean1.getDeclaringBean().getAlternativePriority();
         }
 
-        if (priority2 == null) {
-            return priority1 == null ? 0 : -1;
-        } else if (priority1 == null) {
-            return 1;
+        if (priority2 == null || priority1 == null) {
+            throw new IllegalStateException(String.format(
+                    "Alternative Bean priority should not be null. %s has priority %s; %s has priority %s",
+                    bean1.getBeanClass(), priority1,
+                    bean2.getBeanClass(), priority2));
         }
 
         return priority2.compareTo(priority1);

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/alternatives/priority/AlternativePriorityResolutionTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/alternatives/priority/AlternativePriorityResolutionTest.java
@@ -1,0 +1,83 @@
+package io.quarkus.arc.test.alternatives.priority;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import io.quarkus.arc.AlternativePriority;
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+import javax.annotation.Priority;
+import javax.enterprise.inject.*;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class AlternativePriorityResolutionTest {
+    @RegisterExtension
+    ArcTestContainer testContainer = new ArcTestContainer(NoParentAlternativePriorityProducer1.class,
+            NoParentAlternativePriorityProducer2.class, ParentAlternativePriorityProducer3.class,
+            PrioritizedConsumer.class, MessageBean.class);
+
+    @Test
+    public void testAlternativePriorityResolution() {
+        PrioritizedConsumer bean = Arc.container().instance(PrioritizedConsumer.class).get();
+        assertNotNull(bean, "PrioritizedConsumer bean should not be null");
+        assertEquals("jar", bean.ping());
+    }
+
+    @Singleton
+    static class PrioritizedConsumer {
+        @Inject
+        MessageBean messageBean;
+
+        String ping() {
+            return messageBean.ping();
+        }
+    }
+
+    @Singleton
+    static class NoParentAlternativePriorityProducer1 {
+
+        @Produces
+        @Singleton
+        @AlternativePriority(4)
+        public MessageBean createBar() {
+            return new MessageBean("jar");
+        };
+    }
+
+    @Singleton
+    static class NoParentAlternativePriorityProducer2 {
+
+        @Produces
+        @Singleton
+        @AlternativePriority(2)
+        public MessageBean createBar() {
+            return new MessageBean("far");
+        }
+    }
+
+    @Singleton
+    @Alternative
+    @Priority(2)
+    static class ParentAlternativePriorityProducer3 {
+        @Produces
+        @Singleton
+        public MessageBean createBar() {
+            return new MessageBean("car");
+        }
+    }
+
+    @Vetoed
+    static class MessageBean {
+        final String msg;
+
+        MessageBean(String msg) {
+            this.msg = msg;
+        }
+
+        public String ping() {
+            return msg;
+        }
+    }
+}

